src/shared/rereadpt.c

```diff
@@ -0,0 +1,353 @@
+
+#if HAVE_BLKID
+        r = dlopen_libblkid();
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to load libblkid: %m");
+
+        _cleanup_close_ int lock_fd = -EBADF;
+        if (FLAGS_SET(flags, REREADPT_BSD_LOCK)) {
+                lock_fd = fd_reopen(fd, O_RDONLY|O_CLOEXEC|O_NOCTTY);
+                if (lock_fd < 0)
+                        return log_device_debug_errno(dev, r, "Failed top open lock fd for block device '%s': %m", p);
+
+                if (flock(lock_fd, LOCK_EX|LOCK_NB) < 0)
+                        return log_device_debug_errno(dev, errno, "Failed to take BSD lock on block device '%s': %m", p);
+        }
+
+        _cleanup_(blkid_free_probep) blkid_probe b = NULL;
+        b = sym_blkid_new_probe();
+        if (!b)
+                return log_oom_debug();
+
+        errno = 0;
+        r = sym_blkid_probe_set_device(b, fd, /* off= */ 0, /* size= */ 0);
+        if (r != 0)
+                return log_device_debug_errno(dev, errno_or_else(ENOMEM), "Failed to open block device '%s': %m", p);
+
+        (void) sym_blkid_probe_enable_partitions(b, 1);
+        (void) sym_blkid_probe_set_partitions_flags(b, BLKID_PARTS_ENTRY_DETAILS);
+
+        errno = 0;
+        r = sym_blkid_do_safeprobe(b);
+        if (r == _BLKID_SAFEPROBE_ERROR)
+                return log_device_debug_errno(dev, errno_or_else(EIO), "Unable to probe for partition table of '%s': %m", p);
+        if (IN_SET(r, _BLKID_SAFEPROBE_AMBIGUOUS, _BLKID_SAFEPROBE_NOT_FOUND)) {
+                log_device_debug(dev, "Didn't find partition table on block device '%s', falling back to BLKRRPART.", p);
+                return fallback_ioctl(dev, fd);
+        }
+
+        assert(r == _BLKID_SAFEPROBE_FOUND);
+
+        const char *pttype;
+        (void) sym_blkid_probe_lookup_value(b, "PTTYPE", &pttype, NULL);
+        if (!streq_ptr(pttype, "gpt")) {
+                log_device_debug(dev, "Didn't find a GPT partition table on '%s', falling back to BLKRRPART.", p);
+                return fallback_ioctl(dev, fd);
+        }
+
+        blkid_partlist pl;
+        errno = 0;
+        pl = sym_blkid_probe_get_partitions(b);
+        if (!pl)
+                return log_device_debug_errno(dev, errno_or_else(EIO), "Unable read partition table of '%s': %m", p);
+
+        int n_partitions;
+        errno = 0;
+        n_partitions = sym_blkid_partlist_numof_partitions(pl);
+        if (n_partitions < 0)
+                return log_device_debug_errno(dev, errno_or_else(EIO), "Unable to acquire number of entries in partition table of '%s': %m", p);
+
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        r = partition_enumerator_new(dev, &e);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to enumerate kernel partitions: %m");
+
+        log_debug("Updating/adding kernel partitions...");
```

Comment:
Adding the device path here makes this log message a lot more useful, ditto elsewhere